src/async/withMutex.ts

```diff
@@ -0,0 +1,58 @@
+import { withSemaphore, type SemaphorePermit } from 'radashi'
+
+type AnyFn<T = unknown> = (permit: SemaphorePermit) => Promise<T>
+
+export interface Mutex {
+  isLocked(): boolean
+  acquire(): Promise<SemaphorePermit>
+  release(): void
+}
+
+/**
+ * Creates a mutex-protected instance with supplied function that limits concurrent execution to a single active use.
+ *
+ * @see https://radashi.js.org/reference/async/withSemaphore
+ * @example
+ * ```ts
+ * const limitedFn = withMutex()
+ * limitedFn(() => ...)
+ * ```
+ */
+export function withMutex(): ExclusiveFn
+
+/**
+ * Creates a mutex-protected instance with supplied function that limits concurrent execution to a single active use.
+ * Supports direct invocation and dynamic function passing.
+ *
+ * @see https://radashi.js.org/reference/async/withMutex
+ * @example
+ * ```ts
+ * const limitedFn = withMutex(() => ...)
+ * limitedFn()
+ * limitedFn(() => ...)
```

Comment:
This usage is not supported by the current type definitions. I assume we'll want to update this example, rather than allow overriding the base function?